src/cancelq/posix/cancellable.c

```diff
@@ -0,0 +1,399 @@
+    wrapper.delivery_time = time;
+    if (!heap_insert(&q->heap, &wrapper))
+        goto fail;
+
+    if (HEAP_CHECK)
+        assert(check_queue(q));
+
+    success = true;
+fail:
+    pthread_mutex_unlock(&q->lock);
+done:
+    return success;
+}
+
+void *eq_next_event(event_queue_t *q, abs_time_t time)
+{
+    cancellable_t *next;
+    const void *event;
+
+    event = NULL;
+    next = NULL;
+    if (pthread_mutex_lock(&q->lock) != 0)
+        goto done;
+
+    next = heap_peek(&q->heap, time);
+    if (next == NULL)
+        goto fail;
+
+    event = next->event.event;
+    if (next->id != NOT_CANCELLABLE)
+    {
+        assert(q->ids[next->id] == ID_WAITING);
+        q->ids[next->id] = ID_DELIVERED;
+    }
+
+    heap_rm(&q->heap, 1);
+    if (empty_LH(q))
+        pthread_cond_broadcast(&q->empty);
+
+fail:
+    pthread_mutex_unlock(&q->lock);
+done:
+    return (void *)event;
+}
+
+bool eq_empty(event_queue_t *q)
+{
+    bool empty;
+
+    empty = false;
+    if (pthread_mutex_lock(&q->lock) == 0)
+    {
+        empty = empty_LH(q);
+        pthread_mutex_unlock(&q->lock);
+    }
+
+    return empty;
+}
+
+static cancellation_status_t release_LH(event_queue_t *q, cancellable_id_t id)
+{
+    if (q->idsize < id)
+    {
+        return FAIL_NO_SUCH_ID;
+    }
+    switch (q->ids[id])
+    {
+    case ID_UNUSED:
+        return FAIL_NO_SUCH_ID;
+    case ID_WAITING:
+        return FAIL_NOT_RUN;
+    case ID_DELIVERED:
+        break;
+    }
+    q->ids[id] = ID_UNUSED;
+    return SUCCESS;
+}
+
+static cancellation_status_t cancel_LH(event_queue_t *q, cancellable_id_t id,
+                                       void **event)
+{
+    size_t idx;
+
+    if (q->idsize < id)
+    {
+        return FAIL_NO_SUCH_ID;
+    }
+
+    switch (q->ids[id])
+    {
+    case ID_UNUSED:
+        return FAIL_NO_SUCH_ID;
+    case ID_WAITING:
+        break;
+    case ID_DELIVERED:
+        return FAIL_ALREADY_RUN;
+    }
+
+    // This is pretty deep into the heap's data structure, but heaps
+    // don't do this.
+    for (idx = 1; idx < q->heap.nextidx; idx++)
+    {
+        if (q->heap.data[idx].id == id)
+        {
+            *event = (void *)q->heap.data[idx].event.event;
+            heap_rm(&q->heap, idx);
+            q->ids[id] = ID_UNUSED;
+            return SUCCESS;
+        }
+    }
+    assert(false); // This should never ever happen; it's waiting but corrupted.
+}
+
+cancellation_status_t eq_cancel(event_queue_t *q, cancellable_id_t id,
+                                void **event)
+{
+    cancellation_status_t status;
+
+    status = FAIL_LOCKING;
+    *event = NULL;
+    if (pthread_mutex_lock(&q->lock) == 0)
+    {
+        status = cancel_LH(q, id, event);
+        pthread_mutex_unlock(&q->lock);
+    }
+
+    return status;
+}
+
+cancellation_status_t eq_cancel_or_release(event_queue_t *q,
+                                           cancellable_id_t id,
+                                           void **event)
```

Comment:
I'm not in love with the `event` argument in these functions, but I'm not sure how to handle it.  It's just for memory management, right?  It's times like this that I think we'll need to come up with a general purpose memory manager for the whole run-time.

Reply:
Yeah, it's so the event can be freed if it needs to be.